spec/defines/letsencrypt_certonly_spec.rb

```diff
@@ -207,6 +207,45 @@ class { 'letsencrypt::plugin::dns_cloudflare':
         it { is_expected.to contain_exec('letsencrypt certonly foo.example.com').with_command "letsencrypt --text --agree-tos --non-interactive certonly --rsa-key-size 4096 -a dns-cloudflare --cert-name 'foo.example.com' -d 'foo.example.com' --dns-cloudflare --dns-cloudflare-credentials /etc/letsencrypt/dns-cloudflare.ini --dns-cloudflare-propagation-seconds 10" }
       end
 
+      context 'with dns-gandi plugin' do
+        let(:title) { 'foo.example.com' }
+        let(:params) { { plugin: 'dns-gandi', letsencrypt_command: 'letsencrypt' } }
+        let(:pre_condition) do
+          <<-PUPPET
+          class { 'letsencrypt':
+            email      => 'foo@example.com',
+            config_dir => '/etc/letsencrypt',
+          }
+          class { 'letsencrypt::plugin::dns_gandi':
+            package_name => 'irrelevant',
+            api_key      => 'dummy-gandi-api-token',
+          }
+          PUPPET
+        end
+
+        it { is_expected.to compile.with_all_deps }
+        it { is_expected.to contain_class('letsencrypt::plugin::dns_gandi') }
+        it { is_expected.to contain_exec('letsencrypt certonly foo.example.com').with_command "letsencrypt --text --agree-tos --non-interactive certonly --rsa-key-size 4096 -a dns-gandi --cert-name 'foo.example.com' -d 'foo.example.com' -a certbot-plugin-gandi:dns --certbot-plugin-gandi:dns-credentials /etc/letsencrypt/dns-gandi.ini" }
+      end
+
+      context 'with dns-gandi plugin without apy_key' do
+        let(:title) { 'foo.example.com' }
+        let(:params) { { plugin: 'dns-gandi', letsencrypt_command: 'letsencrypt' } }
+        let(:pre_condition) do
+          <<-PUPPET
+          class { 'letsencrypt':
+            email      => 'foo@example.com',
+            config_dir => '/etc/letsencrypt',
+          }
+          class { 'letsencrypt::plugin::dns_gandi':
+            package_name => 'irrelevant',
+          }
+          PUPPET
+        end
+
+        it { is_expected.not_to compile.with_all_deps }
```

Comment:
Here too it's better to compile and test for a specific error

Reply:
For this one, the module won't work with an empty api_key and I don't know how to handle the error: 
```
  13) letsencrypt::certonly on fedora-36-x86_64 based operating systems with dns-gandi plugin without api_key is expected to fail to compile and raise an error matching /\/expects a value for parameter 'api_key'\//
      Failure/Error: it { is_expected.to compile.and_raise_error(%r{/expects a value for parameter 'api_key'/}) }
        error during compilation: Evaluation Error: Error while evaluating a Resource Statement, Class[Letsencrypt::Plugin::Dns_gandi]: expects a value for parameter 'api_key' (line: 6, column: 11
```
`it { is_expected.to compile.and_raise_error(%r{/expects a value for parameter 'api_key'/}) }` is not correct?

Reply:
The `/` is redundant. In Ruby you can define a regex as `/my regex/` or `%r{my regex}` but `%r{/my regex/}` means you expect `/` in the output. This is useful if you want to test `/var/lib is empty` is in the error. Compare `%r{/var/lib is empty}` to `/\/var\/lib is empty/`.